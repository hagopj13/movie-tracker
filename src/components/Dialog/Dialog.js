// @flow
import React from 'react';
import type { Node } from 'react';
import MuiDialog from '@material-ui/core/Dialog';

import DialogTitle from './Title/DialogTitle';

type Props = {
  isOpen: boolean,
  title: string,
  children: Node,
  onClose: () => void,
  ...
};

const Dialog = (props: Props) => {
  const { isOpen, title, onClose, children, ...dialogProps } = props;

  return (
    <MuiDialog open={isOpen} onClose={onClose} disableRestoreFocus {...(dialogProps: any)}>
      <DialogTitle title={title} onClose={onClose} />
      {children}
    </MuiDialog>
  );
};

export default Dialog;
