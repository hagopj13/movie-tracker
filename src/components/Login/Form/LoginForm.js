// @flow
import React from 'react';
import { useForm } from 'react-hook-form';
import * as yup from 'yup';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';

import PasswordTextField from 'components/Input/PasswordTextField/PasswordTextField';

type Props = {
  isLoginLoading: boolean,
  loginError: string,
  onSubmit: ({ username: string, password: string }) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    '& > *': {
      display: 'block',
      marginBottom: theme.spacing(0.5),
    },
  },
  bottomContainer: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  errorText: {
    marginLeft: theme.spacing(1.5),
  },
}));

const schema = yup.object().shape({
  username: yup.string().required(),
  password: yup.string().required(),
});

const LoginForm = (props: Props) => {
  const { isLoginLoading, loginError, onSubmit } = props;

  const classes = useStyles();
  const { register, handleSubmit, errors } = useForm({
    mode: 'onSubmit',
    validationSchema: schema,
  });

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={classes.root}>
      <TextField
        name="username"
        inputRef={register}
        autoFocus
        fullWidth
        size="small"
        variant="outlined"
        label="Username"
        error={!!errors.username}
        helperText={errors.username?.message ?? ' '}
      />
      <PasswordTextField
        name="password"
        inputRef={register}
        fullWidth
        size="small"
        variant="outlined"
        label="Password"
        error={!!errors.password}
        helperText={errors.password?.message ?? ' '}
      />
      <div className={classes.bottomContainer}>
        <Typography color="error" className={classes.errorText}>
          {loginError}
        </Typography>
        <Button
          type="submit"
          className={classes.submitButton}
          size="large"
          variant="outlined"
          color="primary"
          disabled={isLoginLoading}
        >
          Login
        </Button>
      </div>
    </form>
  );
};

export default LoginForm;
