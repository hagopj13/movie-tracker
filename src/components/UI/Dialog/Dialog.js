import React from 'react';
import MuiDialog from '@material-ui/core/Dialog';

import DialogTitle from './Title/DialogTitle';

type Props = {
  isOpen: boolean,
  title: string,
  children: React.ReactNode,
  onClose: () => void,
};

const Dialog = (props: Props) => {
  const { isOpen, title, onClose, children, ...dialogProps } = props;

  return (
    <MuiDialog open={isOpen} onClose={onClose} disableRestoreFocus {...dialogProps}>
      <DialogTitle onClose={onClose}>{title}</DialogTitle>
      {children}
    </MuiDialog>
  );
};

export default Dialog;
