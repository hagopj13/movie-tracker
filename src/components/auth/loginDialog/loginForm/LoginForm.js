import React from 'react';
import { useForm } from 'react-hook-form';
import * as yup from 'yup';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';

import PasswordTextField from '../../../common/input/passwordTextField/PasswordTextField';

type Props = {
  onSubmit: (data) => void,
};

const useStyles = makeStyles((theme) => ({
  root: {
    '& > *': {
      display: 'block',
      marginBottom: theme.spacing(0.5),
    },
  },
  submitButton: {
    marginLeft: 'auto',
  },
}));

const schema = yup.object().shape({
  username: yup.string().required(),
  password: yup.string().required(),
});

const LoginForm = (props: Props) => {
  const { onSubmit } = props;
  const classes = useStyles();
  const { register, handleSubmit, errors } = useForm({
    mode: 'onSubmit',
    validationSchema: schema,
  });

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={classes.root}>
      <TextField
        name="username"
        inputRef={register}
        autoFocus
        fullWidth
        size="small"
        variant="outlined"
        label="Username"
        error={!!errors.username}
        helperText={errors.username?.message ?? ' '}
      />
      <PasswordTextField
        name="password"
        inputRef={register}
        fullWidth
        size="small"
        variant="outlined"
        label="Password"
        error={!!errors.password}
        helperText={errors.password?.message ?? ' '}
      />
      <Button
        type="submit"
        className={classes.submitButton}
        size="large"
        variant="outlined"
        color="primary"
      >
        Login
      </Button>
    </form>
  );
};

export default LoginForm;
