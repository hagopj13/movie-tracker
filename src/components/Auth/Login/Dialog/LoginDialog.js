import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';

import Dialog from '../../../UI/Dialog/Dialog';
import LoginForm from './Form/LoginForm';
import AuthActionTypes from '../../../../store/auth/auth.types';
import { login, loginClear } from '../../../../store/auth/auth.actions';
import { createIsLoadingSelector } from '../../../../store/api/loading/loading.selectors';
import { createErrorSelector } from '../../../../store/api/error/error.selectors';

type Props = {
  isOpen: boolean,
  isLoginLoading: boolean,
  loginError: string,
  onClose: () => void,
  onLogin: ({ username: string, password: string }) => void,
  onLoginClear: () => void,
};

const useStyles = makeStyles((theme) => ({
  content: {
    padding: theme.spacing(0.5, 5, 1),
    marginBottom: theme.spacing(2),
  },
  contentText: {
    marginBottom: theme.spacing(2),
    maxWidth: '45ch',
  },
}));

const DIALOG_TITLE = 'Login';

const LoginDialog = (props: Props) => {
  const { isOpen, isLoginLoading, loginError, onClose, onLogin, onLoginClear } = props;
  const classes = useStyles();

  const renderInfoText = () => (
    <Typography className={classes.contentText}>
      This app gets its data from the TMDb APIs. To view your account information, login with your
      TMDb credentials in the form below. To create one,&nbsp;
      <Link target="_blank" href="https://www.themoviedb.org/account/signup" color="primary">
        click here
      </Link>
      .
    </Typography>
  );

  return (
    <Dialog isOpen={isOpen} onClose={onClose} title={DIALOG_TITLE} onExited={onLoginClear}>
      <DialogContent className={classes.content}>
        {renderInfoText()}
        <LoginForm isLoginLoading={isLoginLoading} loginError={loginError} onSubmit={onLogin} />
      </DialogContent>
    </Dialog>
  );
};

const mapStateToProps = createStructuredSelector({
  isLoginLoading: createIsLoadingSelector([AuthActionTypes.LOGIN]),
  loginError: createErrorSelector([AuthActionTypes.LOGIN]),
});

const mapDispatchToProps = {
  onLogin: login,
  onLoginClear: loginClear,
};

export default connect(mapStateToProps, mapDispatchToProps)(LoginDialog);
