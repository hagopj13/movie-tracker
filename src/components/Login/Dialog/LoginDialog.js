// @flow
import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';

import Dialog from 'components/Dialog/Dialog';
import AuthActionTypes from 'store/auth/auth.types';
import authActions from 'store/auth/auth.actions';
import loadingSelectors from 'store/api/loading/loading.selectors';
import errorSelectors from 'store/api/error/error.selectors';

import LoginForm from '../Form/LoginForm';

type Props = {
  isOpen: boolean,
  isLoginLoading: boolean,
  loginError: string,
  onClose: () => void,
  onLogin: ({ username: string, password: string }) => void,
  onLoginClear: () => void,
};

const useStyles = makeStyles((theme) => ({
  content: {
    padding: theme.spacing(0.5, 5, 1),
    marginBottom: theme.spacing(2),
  },
  contentText: {
    marginBottom: theme.spacing(2),
    maxWidth: '45ch',
  },
}));

const DIALOG_TITLE = 'Login';

export const LoginDialog = (props: Props) => {
  const { isOpen, isLoginLoading, loginError, onClose, onLogin, onLoginClear } = props;
  const classes = useStyles();

  const renderInfoText = () => (
    <Typography className={classes.contentText}>
      This app gets its data from the TMDb APIs. To view your account information, login with your
      TMDb credentials in the form below. To create one,&nbsp;
      <Link target="_blank" href="https://www.themoviedb.org/account/signup" color="primary">
        click here
      </Link>
      .
    </Typography>
  );

  return (
    <Dialog isOpen={isOpen} onClose={onClose} title={DIALOG_TITLE} onExited={onLoginClear}>
      <DialogContent className={classes.content}>
        {renderInfoText()}
        <LoginForm isLoginLoading={isLoginLoading} loginError={loginError} onSubmit={onLogin} />
      </DialogContent>
    </Dialog>
  );
};

const mapStateToProps = createStructuredSelector({
  isLoginLoading: loadingSelectors.createIsLoadingSelector([AuthActionTypes.LOGIN]),
  loginError: errorSelectors.createErrorSelector([AuthActionTypes.LOGIN]),
});

const mapDispatchToProps = {
  onLogin: authActions.login,
  onLoginClear: authActions.loginClear,
};

export default connect(mapStateToProps, mapDispatchToProps)(LoginDialog);
