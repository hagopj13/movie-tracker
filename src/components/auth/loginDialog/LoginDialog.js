import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { makeStyles } from '@material-ui/core/styles';
import Dialog from '@material-ui/core/Dialog';
import DialogContent from '@material-ui/core/DialogContent';
import Typography from '@material-ui/core/Typography';
import Link from '@material-ui/core/Link';

import DialogTitle from '../../common/dialog/dialogTitle/DialogTitle';
import LoginForm from './loginForm/LoginForm';
import { login, loginClear } from '../../../store/auth/auth.actions';
import { selectIsAuth } from '../../../store/auth/auth.selectors';
import AuthActionTypes from '../../../store/auth/auth.types';
import { createIsLoadingSelector } from '../../../store/api/loading/loading.selectors';
import { createErrorSelector } from '../../../store/api/error/error.selectors';

type Props = {
  isOpen: boolean,
  isAuth: boolean,
  isLoginLoading: boolean,
  loginError: string,
  onClose: () => void,
  onLogin: ({ username: string, password: string }) => void,
  onLoginClear: () => void,
};

const useStyles = makeStyles((theme) => ({
  content: {
    padding: theme.spacing(0.5, 5, 1),
    marginBottom: theme.spacing(2),
  },
  contentText: {
    marginBottom: theme.spacing(2),
    maxWidth: '45ch',
  },
}));

const LoginDialog = (props: Props) => {
  const { isOpen, isAuth, isLoginLoading, loginError, onClose, onLogin, onLoginClear } = props;
  const classes = useStyles();

  useEffect(() => {
    if (isAuth) {
      onClose();
    }
  }, [isAuth, onClose]);

  useEffect(() => {
    if (!isOpen) {
      onLoginClear();
    }
  }, [isOpen, onLoginClear]);

  const renderInfoText = () => (
    <Typography className={classes.contentText}>
      This app gets its data from the TMDb APIs. To view your account information, login with your
      TMDb credentials in the form below. To create one,&nbsp;
      <Link target="_blank" href="https://www.themoviedb.org/account/signup" color="primary">
        click here
      </Link>
      .
    </Typography>
  );

  return (
    <Dialog open={isOpen} onClose={onClose} maxWidth={false}>
      <DialogTitle onClose={onClose}>Login</DialogTitle>
      <DialogContent className={classes.content}>
        {renderInfoText()}
        <LoginForm isLoginLoading={isLoginLoading} loginError={loginError} onSubmit={onLogin} />
      </DialogContent>
    </Dialog>
  );
};

const mapStateToProps = createStructuredSelector({
  isAuth: selectIsAuth,
  isLoginLoading: createIsLoadingSelector([AuthActionTypes.LOGIN]),
  loginError: createErrorSelector([AuthActionTypes.LOGIN]),
});

const mapDispatchToProps = {
  onLogin: login,
  onLoginClear: loginClear,
};

export default connect(mapStateToProps, mapDispatchToProps)(LoginDialog);
