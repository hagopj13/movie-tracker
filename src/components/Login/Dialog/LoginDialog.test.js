import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import Dialog from 'components/Dialog/Dialog';

import { LoginDialog } from './LoginDialog';
import LoginForm from '../Form/LoginForm';

describe('LoginDialog component', () => {
  let shallow;
  const isOpen = true;
  const isLoginLoading = false;
  const loginError = 'some error';
  let mockClose;
  let mockLogin;
  let mockLoginClear;
  let wrapper;

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    mockClose = jest.fn();
    mockLogin = jest.fn();
    mockLoginClear = jest.fn();
    const mockProps = {
      isOpen,
      isLoginLoading,
      loginError,
      onClose: mockClose,
      onLogin: mockLogin,
      onLoginClear: mockLoginClear,
    };
    wrapper = shallow(<LoginDialog {...mockProps} />);
  });

  it('should render the LoginDialog component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass isOpen prop to Dialog', () => {
    expect(wrapper.find(Dialog).prop('isOpen')).toBe(isOpen);
  });

  it('should pass isLoginLoading prop to LoginForm', () => {
    expect(wrapper.find(LoginForm).prop('isLoginLoading')).toBe(isLoginLoading);
  });

  it('should pass loginError prop to LoginForm', () => {
    expect(wrapper.find(LoginForm).prop('loginError')).toBe(loginError);
  });

  it('should call the onClose prop when Dialog is closed', () => {
    wrapper.find(Dialog).prop('onClose')();
    expect(mockClose).toHaveBeenCalledTimes(1);
  });

  it('should call the onLoginClear prop when Dialog exits', () => {
    wrapper.find(Dialog).prop('onExited')();
    expect(mockLoginClear).toHaveBeenCalledTimes(1);
  });

  it('should call the onLogin prop when LoginForm is submitted', () => {
    wrapper.find(LoginForm).prop('onSubmit')();
    expect(mockLogin).toHaveBeenCalledTimes(1);
  });
});
