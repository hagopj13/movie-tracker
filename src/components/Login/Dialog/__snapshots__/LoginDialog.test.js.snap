// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginDialog component should render the LoginDialog component correctly 1`] = `
<Dialog
  isOpen={true}
  onClose={[MockFunction]}
  onExited={[MockFunction]}
  title="Login"
>
  <WithStyles(ForwardRef(DialogContent))
    className="makeStyles-content-1"
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-contentText-2"
    >
      This app gets its data from the TMDb APIs. To view your account information, login with your TMDb credentials in the form below. To create one,Â 
      <WithStyles(ForwardRef(Link))
        color="primary"
        href="https://www.themoviedb.org/account/signup"
        target="_blank"
      >
        click here
      </WithStyles(ForwardRef(Link))>
      .
    </WithStyles(ForwardRef(Typography))>
    <LoginForm
      isLoginLoading={false}
      loginError="some error"
      onSubmit={[MockFunction]}
    />
  </WithStyles(ForwardRef(DialogContent))>
</Dialog>
`;
